src-tauri/src/tools.rs

```diff
@@ -1,22 +1,92 @@
 use crate::globals::{
     get_ip_api_key, get_opencage_key, get_reqwest_client, get_weather_api_user_agent,
 };
+use crate::permissions::{check, Permission, Permission::*};
 use base64::prelude::{Engine as _, BASE64_STANDARD_NO_PAD};
 use chrono::prelude::Local;
 use clipboard::{ClipboardContext, ClipboardProvider};
 use image::{
     codecs::png::PngEncoder, imageops::resize, imageops::FilterType::Triangle, ColorType::Rgba8,
     ImageBuffer, ImageEncoder, Rgba,
 };
+use lazy_static::lazy_static;
 use scrap::{Capturer, Display};
-use serde_json::Value;
+use serde_json::{Map, Value};
 use std::{
-    fs::File, io::ErrorKind::WouldBlock, path::Path, thread::sleep, time::Duration
+    fs::File, future::Future, io::ErrorKind::WouldBlock, path::Path, pin::Pin, sync::Arc, thread::sleep, time::Duration
 };
 use urlencoding::encode;
 
-pub async fn get_location_coordinates(location: &str) -> String {
-    println!("getting {} coordinates!", location);
+type SyncAction = dyn Fn(Map<String, Value>) -> String + Send + Sync;
+type AsyncAction = dyn Fn(Map<String, Value>) -> Pin<Box<dyn Future<Output = String> + Send>> + Send + Sync;
+
+pub enum Action {
+    Sync(Arc<SyncAction>),
+    Async(Arc<AsyncAction>),
+}
+
+pub struct Tool {
+    pub action: Action, // the function that runs when Magnus uses the tool
+    pub description: String, // the message that is displayed on the frontend when Magnus uses the tool
+    pub permissions: Option<Vec<Permission>> // the list of Permissions needed to execute the tool
+}
+
+impl Tool {
+    pub fn new_sync<F>(action: F, description: String, permissions: Option<Vec<Permission>>) -> Self
+    where
+        F: Fn(Map<String, Value>) -> String + Send + Sync + 'static,
+    {
```

Comment:
I need to read more about this dynamic typing stuff (at least thats what I think this is?)  Nice work on all of this!

Reply:
In my understanding, this is just a clean way to denote the type of something if the type is super gnarly and has many traits like we do. I think you could literally just replace F with Fn(Map<String, Value>) -> String + Send + Sync + 'static everywhere and remove the where

Reply:
ohh that makes sense!  So the `String + Send + Sync + 'static` at the end are traits that function being returned inherits? Not sure if that is the correct wording lol

Reply:
yeah I believe the traits are strictly required of the return type of the fn, not the fn itself. so in this case the return type of the fn must be a String that has the Send and Sync traits. the traits may be applying to the fn as a whole?? not sure